_posts/2019-06-07-python-datascience-with-openfaas.md

```diff
@@ -0,0 +1,267 @@
+
+The first thing the Dockerfile does is create a new user and then setup and fix the permissions on some standard folders.  This will ensure that when we change users later, Conda and the function code will run smoothly. The `/opt/conda` folder is often used by Conda and the function code will eventually be put into `/root` (as is done in many of the core templates).
+
+```dockerfile
+RUN addgroup app && adduser app --system --ingroup app \
+    && mkdir -p /opt/conda && chown -R app /opt/conda \
+    && chown -R app /root && chmod -R go+rX /root
+
+ENV HOME /home/app
+ENV PATH=$HOME/conda/bin:$PATH
+
+RUN apt-get update \
+    && apt-get -y install curl bzip2 ${ADDITIONAL_PACKAGE} \
+    && curl -sSL https://repo.continuum.io/miniconda/Miniconda3-latest-Linux-x86_64.sh -o /tmp/miniconda.sh \
+    && curl -sSL https://github.com/openfaas-incubator/of-watchdog/releases/download/0.5.1/of-watchdog > /usr/bin/fwatchdog \
+    && chown app /tmp/miniconda.sh \
+    ...
+```
+
+Once the additional packages and watchdog are installed, we can switch to the non-root `app` user and finish the Python installation
+
+```dockerfile
+WORKDIR /root/
+USER app
+
+RUN bash /tmp/miniconda.sh -bfp $HOME/conda \
+    && conda install -y python=3 \
+    && conda update conda \
+    && conda clean --all --yes \
+    && rm -rf /tmp/miniconda.sh
+```
+
+From here on out, the Docker image will use the `app` user, improving the security of the final function.  This pattern should work for almost any template.  The hard part is knowing which folders the `app` user will need full permissions for. This is where Conda was helpful, because it was easy to find and setup the required permissions and once those folders are configured, Conda behaves as expected when run as a non-root user.
+
+### A multi-module template
+
+Earlier [this year][multifile-blog-post] we highlighted how to use multiple files and modules to organize your Python functions. This new template follows this pattern out of the box.  It provides a "core" module for putting utility methods, a handler file, which is where you need to implement your function logic, and a "training" file where you can put the logic for training your model.
+
+```
+function
+├── __init__.py
+├── core
+│   ├── README.md
+│   ├── __init__.py
+│   └── utils.py
+├── handler.py
+└── train.py
+```
+
+Like the basic `python` template, only the `handler.py` implementation is required, and both the `core` module and the `train.py` could be replaced if you needed, but this structure starts you on a path to keeping the code organized.
+
+## Name classifier implementation
+
```

Comment:
Add some text between headings for formatting or remove a heading.

Reply:
Removed it